pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -466,21 +478,26 @@ def allocate_budget(
         else:
             budgets_size = self.budgets_to_optimize.sum().item()
 
-        # 5. Create an initial guess
-        initial_guess = np.ones(budgets_size) * (total_budget / budgets_size)
-        initial_guess = initial_guess.astype(self._budgets_flat.type.dtype)
-
-        if minimize_kwargs is None:
-            minimize_kwargs = self.DEFAULT_MINIMIZE_KWARGS.copy()
-        else:
-            # Merge with defaults (preferring user-supplied keys)
-            minimize_kwargs = {**self.DEFAULT_MINIMIZE_KWARGS, **minimize_kwargs}
+        # 5. Construct the initial guess (x0) if not provided
+        if x0 is None:
+            x0 = np.ones(budgets_size) * (total_budget / budgets_size).astype(
+                self._budgets_flat.type.dtype
+            )
+        # if x0 arg is provided, validate shape
+        elif x0.shape != (budgets_size,):
+            raise ValueError(
+                f"""The shape of 'x0' {x0.shape} does not match the expected shape {(budgets_size,)}."""
+            )
+        # if x0 arg is provided, validate dtype
+        elif np.issubdtype(x0.dtype, np.number):
```

Comment:
No specifically the `x0.dtype` must match `self._budgets_flat.type.dtype`. I think there's a shortcut you can use for the shape and dtype validation: `x0 = self.budget_flat.type.filter(x0)`, which will do any allowed casting or raise if things cannot be safely cast (and also checks shape)